docs/ELF_DLOPEN_METADATA.md

```diff
@@ -0,0 +1,84 @@
+providing a standardized specification for a custom ELF note that can be used to list `dlopen()`
+dependencies.
+
+## Implementation
+
+This document will attempt to define a common metadata format specification, so that multiple implementers
+might use it when coding upstream software, and packagers might use it when building packages and setting
+dependencies.
+
+The metadata will be embedded in a series of new, 4-bytes-aligned, allocated, 0-padded, read-only ELF header
+sections, in a name-value JSON object format, either one ELF note per dependency or as a single note using a
+json list. Implementers working on parsing ELF files should not assume a specific list of names, but parse
+anything that is included in the section, and should look for the note using the `note type`. Implementers
+working on build tools should strive to use the same names, for consistency. The most common will be listed
+here.
+
+* Section header
+
+```
+SECTION: `.note.dlopen`
+note type: `0x407c0c0a`
+Owner: `FDO` (FreeDesktop.org)
+Value: a single JSON object encoded as a zero-terminated UTF-8 string
+```
+
+* JSON payload
+
+```json
+{
+     "soname":      "libfoo.so.1",
+     "feature":     "foo",
+     "description": "Enables the foo feature",
+     "priority":    "recommended"
+}
+```
+
+The format is a single JSON object, encoded as a zero-terminated `UTF-8` string. Each key in the object
+shall be unique as per recommendations of [RFC8259](https://datatracker.ietf.org/doc/html/rfc8259#section-4).
+Strings shall not contain any control character, nor use `\uXXX` escaping.
+
+Reference implementations of [packaging tools for .deb and .rpm](https://github.com/systemd/package-notes)
+are available, and provide macros/helpers to parse the note when building packages and adding dependencies.
+
+## Well-known keys
+
+The metadata format is intentionally extensible, so that upstreams and later revisions of this spec can add
+their own information. The 'soname' field is required, everything else is optional. If alternative soname
+versions for the same library are supported at the same time, an array can be used, listing the most
+preferred first, and it's up to the parser implementation to choose whether to use all of them or not. If
+the priority field is used, it must follow the specification and use one of the values specified in the
+table. If it is not specified, a parser should assume 'recommended' if a priority is needed.
+
+| Key name    | Key type                   | Mandatory | Key description                                                          | Example value                         |
+|-------------|----------------------------|-----------|--------------------------------------------------------------------------|---------------------------------------|
+| soname      | string or array of strings | yes       | The library name(s) as loaded by `dlopen()`                              | [ "libfoo.so.1", "libfoo.so.0" ]      |
```

Comment:
I'd just make it an array unconditionally. Having type variants complicates parsing code…

Please don't use "(s)", it's distracting and visually displeasing. Just "The library names loaded by `dlopen()`" works better.